text/0000-standard-lazy-types.md

```diff
@@ -0,0 +1,457 @@
+    pub fn get(&self) -> Option<&T>;
+
+    /// Sets the contents of this cell to `value`.
+    ///
+    /// Returns `Ok(())` if the cell was empty and `Err(value)` if it was
+    /// full.
+    pub fn set(&self, value: T) -> Result<(), T>;
+
+    /// Gets the contents of the cell, initializing it with `f`
+    /// if the cell was empty.
+    ///
+    /// # Panics
+    ///
+    /// If `f` panics, the panic is propagated to the caller, and the cell
+    /// remains uninitialized.
+    ///
+    /// It is an error to reentrantly initialize the cell from `f`. Doing
+    /// so results in a panic.
+    pub fn get_or_init<F>(&self, f: F) -> &T
+    where
+        F: FnOnce() -> T,
+    ;
+
+    /// Gets the contents of the cell, initializing it with `f` if
+    /// the cell was empty. If the cell was empty and `f` failed, an
+    /// error is returned.
+    ///
+    /// # Panics
+    ///
+    /// If `f` panics, the panic is propagated to the caller, and the cell
+    /// remains uninitialized.
+    ///
+    /// It is an error to reentrantly initialize the cell from `f`. Doing
+    /// so results in a panic.
+    pub fn get_or_try_init<F, E>(&self, f: F) -> Result<&T, E>
+    where
+        F: FnOnce() -> Result<T, E>,
+    ;
+}
+```
+
+Notable features of the API:
+
+* `OnceCell` is created empty, by a const fn.
+* Initialization succeeds at most once.
+* `get_or_init` and `get_or_try_init` methods can be used to conveniently initialize a cell.
+* `get_` family of methods return `&T`.
+
+Similarly to other interior mutability primitives, `OnceCell` comes in two flavors:
+
+* Non thread-safe `std::cell::OnceCell`.
+* Thread-safe `std::sync::OnceCell`.
+
+Here's how `OnceCell` can be used to implement lazy-initialized global data:
+
+```rust
+use std::{sync::{Mutex, OnceCell}, collections::HashMap};
+
+fn global_data() -> &'static Mutex<HashMap<i32, String>> {
+    static INSTANCE: OnceCell<Mutex<HashMap<i32, String>>> = OnceCell::new();
+    INSTANCE.get_or_init(|| {
+        let mut m = HashMap::new();
+        m.insert(13, "Spica".to_string());
+        m.insert(74, "Hoyten".to_string());
+        Mutex::new(m)
+    })
+}
+```
+
+Here's how `OnceCell` can be used to implement a lazy field:
+
+```rust
+use std::{fs, io, path::PathBuf, cell::OnceCell};
+
+struct Ctx {
+    config_path: PathBuf,
+    config: OnceCell<String>,
+}
+
+impl Ctx {
+    pub fn get_config(&self) -> Result<&str, io::Error> {
+        let cfg = self.config.get_or_try_init(|| {
+            fs::read_to_string(&self.config_path)
+        })?;
+        Ok(cfg.as_str())
+    }
+}
+```
+
+We also provide a more convenient but less powerful `Lazy<T, F>` wrapper around `OnceCell<T>`, which allows to specify the initializing closure at creation time:
+
+```rust
+pub struct Lazy<T, F = fn() -> T> { ... }
+
+impl<T, F> Lazy<T, F> {
+    /// Creates a new lazy value with the given initializing function.
+    pub const fn new(init: F) -> Lazy<T, F>;
```

Comment:
I guess we can. Not sure how useful it would be. In particular, we can't spell `const impl From` yet. As adding `From` is backwards compatible, I suggest omitting this from the initial implementation.